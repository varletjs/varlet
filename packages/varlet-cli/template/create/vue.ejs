<template>
  <div class="var-<%- kebabCaseName %>"></div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'
import './<%- kebabCaseName %>.less'

export default defineComponent({
  name: '<%- namespace + bigCamelizeName %>',
})
</script>
