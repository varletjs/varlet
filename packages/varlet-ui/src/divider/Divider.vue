<template>
  <div
    class="var-divider var--box"
    :class="[
      vertical ? 'var-divider--vertical' : null,
      withText ? 'var-divider--with-text' : null,
      isInset ? 'var-divider--inset' : null,
      dashed ? 'var-divider--dashed' : null,
    ]"
    :style="style"
  >
    <slot>
      <span class="var-divider__text" v-if="description">{{ description }}</span>
    </slot>
  </div>
</template>

<script lang="ts">
import { defineComponent, computed, reactive, onMounted, onUpdated, toRefs } from 'vue'
import { toSizeUnit } from '../utils/elements'
import { isBool, toNumber } from '../utils/shared'
import { props } from './props'

export default defineComponent({
  name: 'VarDivider',
  props,
  setup(props, { slots }) {
    const state = reactive({ withText: false })

    const isInset = computed(() => (isBool(props.inset) ? props.inset : true))

    const style = computed(() => {
      const { inset, vertical, margin } = props
      const baseStyle = {
        margin,
      }

      if (isBool(inset) || inset === 0) return { ...baseStyle }

      // -18px -> -18
      const _inset = toNumber(inset)
      // -18px -> 18px
      const absInsetWithUnit = Math.abs(_inset) + (inset + '').replace(_inset + '', '')
      return vertical
        ? {
            ...baseStyle,
            height: `calc(80% - ${toSizeUnit(absInsetWithUnit)})`,
          }
        : {
            ...baseStyle,
            width: `calc(100% - ${toSizeUnit(absInsetWithUnit)})`,
            left: _inset > 0 ? toSizeUnit(absInsetWithUnit) : toSizeUnit(0),
          }
    })

    const checkHasText = () => {
      state.withText = Boolean(slots.default) || Boolean(props.description)
    }

    onMounted(() => {
      checkHasText()
    })

    onUpdated(() => {
      checkHasText()
    })

    return {
      ...toRefs(state),
      style,
      isInset,
    }
  },
})
</script>

<style lang="less">
@import '../styles/common';
@import './divider';
</style>
