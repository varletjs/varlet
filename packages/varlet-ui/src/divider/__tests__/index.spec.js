import VarDivider from '../Divider'
import Divider from '..'
import { createApp } from 'vue'
import { mount } from '@vue/test-utils'

test('test divider plugin', () => {
  const app = createApp({}).use(Divider)
  expect(app.component(Divider.name)).toBeTruthy()
})

describe('test divider components props', () => {
  test('test divider inset', async () => {
    const wrapper = mount(VarDivider, {
      props: {
        inset: true,
      },
    })

    expect(wrapper.find('.var-divider--inset').exists()).toBe(true)
    await wrapper.setProps({ inset: false })
    expect(wrapper.find('.var-divider--inset').exists()).toBe(false)
    wrapper.unmount()
  })

  test('test divider vertical', async () => {
    const wrapper = mount(VarDivider, {
      props: {
        vertical: true,
      },
    })

    expect(wrapper.find('.var-divider--vertical').exists()).toBe(true)
    await wrapper.setProps({ vertical: false })
    expect(wrapper.find('.var-divider--vertical').exists()).toBe(false)
    wrapper.unmount()
  })

  test('test divider dashed', async () => {
    const wrapper = mount(VarDivider, {
      props: {
        dashed: true,
      },
    })

    expect(wrapper.find('.var-divider--dashed').exists()).toBe(true)
    await wrapper.setProps({ dashed: false })
    expect(wrapper.find('.var-divider--dashed').exists()).toBe(false)
    wrapper.unmount()
  })

  test('test divider description', () => {
    const wrapper = mount(VarDivider, {
      props: {
        description: 'This is description',
      },
    })

    expect(wrapper.find('.var-divider__text').element.textContent).toBe('This is description')
    wrapper.unmount()
  })

  test('test divider margin', () => {
    const wrapper = mount(VarDivider, {
      props: {
        margin: '20px',
      },
    })

    expect(wrapper.find('.var-divider').attributes('style')).toContain('margin: 20px;')
    wrapper.unmount()
  })
})
