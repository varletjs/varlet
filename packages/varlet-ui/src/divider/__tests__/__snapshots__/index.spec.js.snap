// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test default slot of divider 1`] = `"<div class=\\"var-divider var--box\\"><span>description</span></div>"`;

exports[`test divider dashed and margin prop 1`] = `
"<div class=\\"var-divider var--box\\">
  <!--v-if-->
</div>"
`;

exports[`test divider dashed and margin prop 2`] = `
"<div class=\\"var-divider var--box var-divider--dashed\\">
  <!--v-if-->
</div>"
`;

exports[`test divider dashed and margin prop 3`] = `
"<div class=\\"var-divider var--box\\" style=\\"margin: 36px 0px;\\">
  <!--v-if-->
</div>"
`;

exports[`test divider example 1`] = `
"<div class=\\"app-type\\">基本使用</div>
<div class=\\"var-divider var--box\\">
  <!--v-if-->
</div>
<div class=\\"app-type\\">虚线</div>
<div class=\\"var-divider var--box var-divider--dashed\\">
  <!--v-if-->
</div>
<div class=\\"app-type\\">缩进</div>
<div class=\\"var-divider var--box var-divider--inset\\">
  <!--v-if-->
</div>
<div class=\\"var-divider var--box var-divider--inset\\" style=\\"margin: 36px 0px; width: calc(100% - 36px); left: 36px;\\">
  <!--v-if-->
</div>
<div class=\\"var-divider var--box var-divider--inset\\" style=\\"width: calc(100% - 36px); left: 0px;\\">
  <!--v-if-->
</div>
<div class=\\"app-type\\">垂直分割线</div>
<div class=\\"divider-example-vertical-container\\"><span>文字</span>
  <div class=\\"var-divider var--box var-divider--vertical\\">
    <!--v-if-->
  </div><span>文字</span>
  <div class=\\"var-divider var--box var-divider--vertical\\">
    <!--v-if-->
  </div><span>文字</span>
</div>
<div class=\\"app-type\\">带有文字描述的分割线</div>
<div class=\\"var-divider var--box\\"><span class=\\"var-divider__text\\">文字描述</span></div>
<div class=\\"app-type\\">自定义</div>
<div class=\\"var-divider var--box\\"><i class=\\"var-icon var-icon--set var-icon-heart-outline\\" style=\\"color: rgb(41, 121, 255); transition: transform 0ms; margin: 0px 16px;\\"></i></div>"
`;

exports[`test divider inset and vertical prop 1`] = `
"<div class=\\"var-divider var--box var-divider--inset\\">
  <!--v-if-->
</div>"
`;

exports[`test divider inset and vertical prop 2`] = `
"<div class=\\"var-divider var--box var-divider--inset\\" style=\\"width: calc(100% - 36px); left: 36px;\\">
  <!--v-if-->
</div>"
`;

exports[`test divider inset and vertical prop 3`] = `
"<div class=\\"var-divider var--box var-divider--inset\\" style=\\"width: calc(100% - 36px); left: 0px;\\">
  <!--v-if-->
</div>"
`;

exports[`test divider inset and vertical prop 4`] = `
"<div class=\\"var-divider var--box var-divider--vertical var-divider--inset\\" style=\\"height: calc(80% - 36px);\\">
  <!--v-if-->
</div>"
`;
